Configure memcheck tools

jamiesnape opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3920)

<!-- Reviewable:end -->

jamiesnape commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

jamiesnape commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

jamiesnape commented:
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please

jamiesnape commented:
@drake-jenkins-bot retest this please
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-debug please
@drake-jenkins-bot linux-clang-ninja-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please
@drake-jenkins-bot linux-gcc-ninja-experimental-open-source-debug please

jamiesnape commented:
Depends on #4135.

jamiesnape commented:
@drake-jenkins-bot linux-clang-experimental please

jamiesnape commented:
@david-german-tri This is fundamentally incompatible with Bazel. How do you wish to proceed?

david-german-tri commented:
> This is fundamentally incompatible with Bazel.

I see that this is not useful for `bazel test` targets, of course.  I also see that the compiler and linker flags this PR sets will not propagate through shell-outs to `bazel build`.  However, I would think this PR still provides users a path to run CTests under valgrind, even if they happen to be linked against some `.a`s built in Bazel, right?  Similarly, it provides a path to link any code built from CMake against `san` libraries.

> How do you wish to proceed?

How much more work is there to make this PR ready?  If it's reasonably close, I'd like to proceed, because it's certainly the fastest way to resolving #3873 and locking in our memory-sanity.   It will also provide a benchmark for future bazel check/san configs.

jamiesnape commented:
> How much more work is there to make this PR ready?

Testing and adding examples of the output to the documentation. No more than a day. It should be usable now, though the sanitizer builds may fail due to there being issues with some of the externals.

david-german-tri commented:
Nice!  Go for it, then.

jamiesnape commented:
+@mwoehlke-kitware for feature review and +@david-german-tri for platform review.

mwoehlke-kitware commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[cmake/options.cmake, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYymPK3zwzkIbMrI_e-:-KYymPK4JokJ5B2XKCe7:bi5nt3s) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/options.cmake#L194)):*
> ```CMake
>   drake_system_dependency(
>     LCM OPTIONAL REQUIRES lcm
>     DEPENDS "NOT USE_SANITIZER"
> ```

BTW, I thought we fixed `lcm-gen` to remove sanitizer flags? What's the problem here, and can/should it be fixed in LCM instead?

---

*[cmake/tools.cmake, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYynVoTLQwi4zrL9oQD:-KYynVoUNFLfAelLHsHK:b-lfgo5c) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L72)):*
> ```CMake
> # flag.
> #------------------------------------------------------------------------------
> macro(drake_add_compiler_and_linker_flag FLAG NAME)
> ```

BTW do we really need `NAME` here? Can this either be a function or set/unset some scratch variable to test for flag support, instead?

---

*[cmake/tools.cmake, line 212 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzGLf5RsvTuGrX2Woy:-KYzGLf5RsvTuGrX2Woz:b-34bbxh) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L212)):*
> ```CMake
>       set(ENV{LSAN_OPTIONS}
>         "${SANITIZER_COMMON_OPTIONS}:suppressions=${_tools_dir}/lsan.supp:$ENV{LSAN_OPTIONS}")
>     elseif(USE_SANITIZER MATCHES "Memory(WithOrigins)?")
> ```

`^` and `$`?

---

*[cmake/tools.cmake, line 272 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzGf-Z7uBoVx9QBBeY:-KYzGf-Z7uBoVx9QBBeZ:bc0guwy) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L272)):*
> ```CMake
>         set(_trace_children_skip "/bin/*,/usr/bin/*,/usr/local/bin/*")
> 
>         if(Matlab_FOUND AND NOT DISABLE_MATLAB)
> ```

Um... per above, isn't `DISABLE_MATLAB=ON`, always?

---

*[drake/doc/dynamic_analysis_tools.rst, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzIj70_DiCXCRi-Duw:-KYzIj71_mj9CW-089-p:bulotzp) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L21)):*
> ```reStructuredText
> Python are disabled.
> 
> .. _dynamic_analysis_tools_sanitizers:
> ```

BTW, s.b. `_dynamic_analysis_tools-sanitizers`? (`s/_/-/`)

---

*[drake/doc/dynamic_analysis_tools.rst, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzIx4E8SW7GeDhmENm:-KYzIx4FH6H2SMXp_xpv:b-l71y9y) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L36)):*
> ```reStructuredText
> 
>     mkdir build && cd build
>     cmake -DCMAKE_BUILD_TYPE:STRING=Debug -DUSE_SANITIZER:BOOL=Address ..
> ```

...pretty sure you meant `s/BOOL/STRING/` (multiple instances).

Also, as a consistency note, you give types here, but not in the valgrind section... Omitting the type for `CMAKE_BUILD_TYPE` may actually be better, as otherwise CMake does not fill in the documentation string for this variable.

---

*[drake/doc/dynamic_analysis_tools.rst, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzJt0212D_ORQkegz8:-KYzJt0212D_ORQkegz9:b-e9lc45) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L102)):*
> ```reStructuredText
> .. _dynamic_analysis_tools-valgrind_tools:
> 
> Valgrind Tools
> ```

BTW, there seems to be support for many more valgrind tools than are documented?

---

*[drake/doc/lint_static_analysis_tools.rst, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMBzjhDYyYXwu8RJZ:-KYzMBzkPpraRVxjgcs4:b-2d3y8z) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L11)):*
> ```reStructuredText
>    :local:
> 
> .. _static_analysis_tools-introduction:
> ```

bookmark name is inconsistent

---

*[drake/doc/lint_static_analysis_tools.rst, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMQpEdhJCzn-mJ6Jf:-KYzMQpEdhJCzn-mJ6Jg:bxy6r30) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L37)):*
> ```reStructuredText
> 
> `include-what-you-use <http://include-what-you-use.org>`_ is a tool based on
> Clang to identify include-what-you-use violations . ::
> ```

BTW, can you elaborate on this? "X is a tool to detect violations of X" is rather more tautological than informative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
I'll hold review until feature lgtm

---

Review status: 0 of 10 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-:-KZ48HIHxEUt8KQHLVsu:b-s33vr0)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[cmake/options.cmake, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYymPK3zwzkIbMrI_e-:-KZ-pk7lxGVHrahxIk8A:b-54fepg) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/options.cmake#L194)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW, I thought we fixed `lcm-gen` to remove sanitizer flags? What's the problem here, and can/should it be fixed in LCM instead?
</blockquote></details>

BTW Apparently the fix did not work. Hopefully, we can add some pragmas to LCM.

---

*[cmake/tools.cmake, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYynVoTLQwi4zrL9oQD:-KZ70ZuIvLiMJG1IAIkU:bykf8jb) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L72)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW do we really need `NAME` here? Can this either be a function or set/unset some scratch variable to test for flag support, instead?
</blockquote></details>

BTW It just feels neater and it is what LLVM themselves do.

---

*[cmake/tools.cmake, line 212 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzGLf5RsvTuGrX2Woy:-KZ70itY-zBiOa7qQhpW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L212)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

`^` and `$`?
</blockquote></details>

Done.

---

*[cmake/tools.cmake, line 272 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzGf-Z7uBoVx9QBBeY:-KZ70k03Q2jLk1qbsq7_:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/tools.cmake#L272)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Um... per above, isn't `DISABLE_MATLAB=ON`, always?
</blockquote></details>

Removed.

---

*[drake/doc/dynamic_analysis_tools.rst, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzIj70_DiCXCRi-Duw:-KZ70lyQatUgQQHauZLC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L21)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW, s.b. `_dynamic_analysis_tools-sanitizers`? (`s/_/-/`)
</blockquote></details>

Done.

---

*[drake/doc/dynamic_analysis_tools.rst, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzIx4E8SW7GeDhmENm:-KZ70mjcZxknSRVHN18S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L36)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

...pretty sure you meant `s/BOOL/STRING/` (multiple instances).

Also, as a consistency note, you give types here, but not in the valgrind section... Omitting the type for `CMAKE_BUILD_TYPE` may actually be better, as otherwise CMake does not fill in the documentation string for this variable.
</blockquote></details>

Done.

---

*[drake/doc/dynamic_analysis_tools.rst, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzJt0212D_ORQkegz8:-KZ70pT_hJkdFKE6R_tI:b-mr93jx) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/dynamic_analysis_tools.rst#L102)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW, there seems to be support for many more valgrind tools than are documented?
</blockquote></details>

BTW True. I have not tested all them, though.

---

*[drake/doc/lint_static_analysis_tools.rst, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMBzjhDYyYXwu8RJZ:-KZ712RC6Hj1Wo_Qnxzr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L11)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

bookmark name is inconsistent
</blockquote></details>

Done.

---

*[drake/doc/lint_static_analysis_tools.rst, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMQpEdhJCzn-mJ6Jf:-KZ716Fffevop3du3Bn7:b-d3nj2d) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L37)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

BTW, can you elaborate on this? "X is a tool to detect violations of X" is rather more tautological than informative.
</blockquote></details>

BTW Not sure I have a succinct  alternative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[cmake/options.cmake, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYymPK3zwzkIbMrI_e-:-KZ7I3-PqImzMKRaGzaK:b9d3ns1) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/options.cmake#L194)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

BTW Apparently the fix did not work. Hopefully, we can add some pragmas to LCM.
</blockquote></details>

Is there an issue open for this?

---

*[drake/doc/lint_static_analysis_tools.rst, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMQpEdhJCzn-mJ6Jf:-KZ7HDXUUE7GsbU7i562:b-l2lwna) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L37)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

BTW Not sure I have a succinct  alternative.
</blockquote></details>

"...to identify unnecessary and/or missing \`#include\`s."?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
> I'll hold review until feature lgtm

FYI, LGTM except for the IWYU blurb.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-:-KZ7IS1rlvdZEs3kVV6V:b-5w3iv4)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[cmake/options.cmake, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYymPK3zwzkIbMrI_e-:-KZ7JrAq8a_yDo07QUNX:bsxqrpj) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/cmake/options.cmake#L194)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Is there an issue open for this?
</blockquote></details>

I guess it is an LCM issue really, so I will leave it to you to interact with them.

---

*[drake/doc/lint_static_analysis_tools.rst, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KYzMQpEdhJCzn-mJ6Jf:-KZ7KA3aI106yy_7Gpwg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c06153149d099e4783841c1a06a9ead245e5d35c/drake/doc/lint_static_analysis_tools.rst#L37)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

"...to identify unnecessary and/or missing \`#include\`s."?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-:-KZ7MUvFS7rvInHqqtlp:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
PTAL @david-german-tri.

david-german-tri commented:
Looking now.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-:-KZN74FGNgEioLyU59Cn:bcc08gq)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 3 of 11 files at r1, 2 of 4 files at r2.
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNCc5b9pAKJhmyQXID:bwvue1e) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
> ```CMake
> # supports the flag.
> #------------------------------------------------------------------------------
> macro(drake_add_compiler_flag_if_supported FLAG NAME)
> ```

Is there a CMake code convention for private helper functions?  I don't think we want anyone to call this (or the other helpers) from other files.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNBD6JhWtivuz1C4vM:b-s8ijt9) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
> ```CMake
> macro(drake_add_compiler_and_linker_flag FLAG NAME)
>   set(SAFE_CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES})
>   set(CMAKE_REQUIRED_LIBRARIES ${FLAG})
> ```

Is this dead code?  I'm kind of confused by all three `set` statements in this block, actually.  Seems like they do nothing on net.

---

*[cmake/tools.cmake, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBpKSS3eyhWXl-zRX:-KZNBpKSS3eyhWXl-zRY:b-3pyo05) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L89)):*
> ```CMake
>     message(STATUS
>       "Disabling Fortran because compiler does not support ${FLAG}")
>     set(DISABLE_FORTRAN ON)
> ```

This seems guaranteed to surprise someone at some point.  Can we just error helpfully instead?

---

*[cmake/tools.cmake, line 280 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZND7GH1YEVBZQnvPIa:-KZND7GH1YEVBZQnvPIb:b-nwj1nz) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L280)):*
> ```CMake
>         endif()
> 
>         # TODO(jamiesnape): Use common Valgrind suppressions for Bazel and CMake.
> ```

FYI I've added this to #3873.

---

*[drake/doc/dynamic_analysis_tools.rst, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFUH0HBywIIOQ5gYu:-KZNFUH0HBywIIOQ5gYv:bzcgny5) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L39)):*
> ```reStructuredText
>     cmake --build .
>     cd drake
>     ctest --extra-verbose --test-action MemCheck .
> ```

This fails for me with:

```
$ ctest --extra-verbose --test-action MemCheck .
UpdateCTestConfiguration  from :/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Parse Config file:/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
   Site: dgerman01
   Build name: Linux-g++-4.9
UpdateCTestConfiguration  from :/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Parse Config file:/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Memory check project /home/dgerman/drake/3920/drake/build/drake
Memory checker (MemoryCheckCommand) not set, or cannot find the specified program.
Errors while running CTest
```

But if I just `ctest -VV -R diagram_test`, I get perfectly good ASAN failures (having hacked up the test to have a memory violation).

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZNFus7F5wt3kWBsnHy:b-v7suxr) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
> ```reStructuredText
> 
>     mkdir build && cd build
>     cmake -DCMAKE_BUILD_TYPE=Debug -DUSE_SANITIZER=Memory ..
> ```

This fails on gcc (as it should), but worth mentioning here.  Also, the error message includes some surprising fortran-related goop on the first attempt to run this command, which goes away on the second attempt.

```
-- Performing Test Fortran_COMPILER_SUPPORTS_FSANITIZE_MEMORY
CMake Error at /home/dgerman/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CheckFortranSourceCompiles.cmake:76 (try_compile):
  Unknown extension ".F" for file

    /home/dgerman/drake/3920/drake/build/CMakeFiles/CMakeTmp/src.F

  try_compile() works only for enabled languages.  Currently these are:

    C CXX

  See project() command to enable other languages.
Call Stack (most recent call first):
  /home/dgerman/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CheckFortranCompilerFlag.cmake:53 (CHECK_Fortran_SOURCE_COMPILES)
  cmake/tools.cmake:58 (check_fortran_compiler_flag)
  cmake/tools.cmake:75 (drake_add_compiler_flag_if_supported)
  cmake/tools.cmake:213 (drake_add_compiler_and_linker_flag)
  cmake/tools.cmake:35 (drake_setup_sanitizers)
  cmake/tools.cmake:13 (drake_setup_tools_common)
  CMakeLists.txt:15 (drake_setup_tools_superbuild)


-- Performing Test Fortran_COMPILER_SUPPORTS_FSANITIZE_MEMORY - Failed
CMake Error at cmake/tools.cmake:79 (message):
  C++ compiler does not support -fsanitize=memory
Call Stack (most recent call first):
  cmake/tools.cmake:213 (drake_add_compiler_and_linker_flag)
  cmake/tools.cmake:35 (drake_setup_sanitizers)
  cmake/tools.cmake:13 (drake_setup_tools_common)
  CMakeLists.txt:15 (drake_setup_tools_superbuild)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNJEL3lg9yVm_AV_gP:b2nhky0) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Is there a CMake code convention for private helper functions?  I don't think we want anyone to call this (or the other helpers) from other files.
</blockquote></details>

No.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNHG1Clpv9woOYe6h5:b-ved04p) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Is this dead code?  I'm kind of confused by all three `set` statements in this block, actually.  Seems like they do nothing on net.
</blockquote></details>

No, it is probably the most important line in the function. It the required flag to be tested and you do not want it left in CMAKE_REQUIRED_LIBRARIES if it fails.

---

*[cmake/tools.cmake, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBpKSS3eyhWXl-zRX:-KZNHbeu6C-QtVAtFW3B:b-pcojcr) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L89)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This seems guaranteed to surprise someone at some point.  Can we just error helpfully instead?
</blockquote></details>

Disabling Fortran is the desired result here. It is not an error.

---

*[drake/doc/dynamic_analysis_tools.rst, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFUH0HBywIIOQ5gYu:-KZNIETZ8fOLPCWI-58t:bk2jaso) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L39)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This fails for me with:

```
$ ctest --extra-verbose --test-action MemCheck .
UpdateCTestConfiguration  from :/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Parse Config file:/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
   Site: dgerman01
   Build name: Linux-g++-4.9
UpdateCTestConfiguration  from :/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Parse Config file:/home/dgerman/drake/3920/drake/build/drake/DartConfiguration.tcl
Memory check project /home/dgerman/drake/3920/drake/build/drake
Memory checker (MemoryCheckCommand) not set, or cannot find the specified program.
Errors while running CTest
```

But if I just `ctest -VV -R diagram_test`, I get perfectly good ASAN failures (having hacked up the test to have a memory violation).

</blockquote></details>

Which tool? If it is anything other than Valgrind, you do not need the MemCheck.

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZNIgbfhabxElLZNih0:b-s0tlou) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This fails on gcc (as it should), but worth mentioning here.  Also, the error message includes some surprising fortran-related goop on the first attempt to run this command, which goes away on the second attempt.

```
-- Performing Test Fortran_COMPILER_SUPPORTS_FSANITIZE_MEMORY
CMake Error at /home/dgerman/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CheckFortranSourceCompiles.cmake:76 (try_compile):
  Unknown extension ".F" for file

    /home/dgerman/drake/3920/drake/build/CMakeFiles/CMakeTmp/src.F

  try_compile() works only for enabled languages.  Currently these are:

    C CXX

  See project() command to enable other languages.
Call Stack (most recent call first):
  /home/dgerman/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CheckFortranCompilerFlag.cmake:53 (CHECK_Fortran_SOURCE_COMPILES)
  cmake/tools.cmake:58 (check_fortran_compiler_flag)
  cmake/tools.cmake:75 (drake_add_compiler_flag_if_supported)
  cmake/tools.cmake:213 (drake_add_compiler_and_linker_flag)
  cmake/tools.cmake:35 (drake_setup_sanitizers)
  cmake/tools.cmake:13 (drake_setup_tools_common)
  CMakeLists.txt:15 (drake_setup_tools_superbuild)


-- Performing Test Fortran_COMPILER_SUPPORTS_FSANITIZE_MEMORY - Failed
CMake Error at cmake/tools.cmake:79 (message):
  C++ compiler does not support -fsanitize=memory
Call Stack (most recent call first):
  cmake/tools.cmake:213 (drake_add_compiler_and_linker_flag)
  cmake/tools.cmake:35 (drake_setup_sanitizers)
  cmake/tools.cmake:13 (drake_setup_tools_common)
  CMakeLists.txt:15 (drake_setup_tools_superbuild)
```

</blockquote></details>

I will look into it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNJRsnIxJlzDy69Fca:b-wic6w2) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

No.
</blockquote></details>

Private stuff (macros/functions, but also local variables) can be prefixed with `_`. I've used this in LCM, and it is also seen in `ExternalProject.cmake`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 4 of 11 files at r1.
Review status: 8 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNL_Wl88LotI8wvCUe:b-p2niey) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
> can be prefixed with _.

FWIW I like this idea.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNLoJKsVcmK576dQrC:b-kj6ev3) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

No, it is probably the most important line in the function. It the required flag to be tested and you do not want it left in CMAKE_REQUIRED_LIBRARIES if it fails.
</blockquote></details>

You're the expert, but I really don't get it.  How is this not:

```
b = a;
a = c;
fn(c, d);
a = b;
```

---

*[cmake/tools.cmake, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBpKSS3eyhWXl-zRX:-KZNMhjIEajmxX8urGk1:b-megd8p) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L89)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Disabling Fortran is the desired result here. It is not an error.
</blockquote></details>

OK, I guess that makes sense, thinking of this function as private.

---

*[drake/doc/dynamic_analysis_tools.rst, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFUH0HBywIIOQ5gYu:-KZNL5siqhl82p-VaDsY:b-1lep51) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L39)):*
AddressSanitizer (which is why I'm commenting here).

>  If it is anything other than Valgrind, you do not need the MemCheck.

Sounds like the documentation should be corrected here and below, then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 8 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNKC-DvfeWeoUkwdRI:bnm1dgo) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Private stuff (macros/functions, but also local variables) can be prefixed with `_`. I've used this in LCM, and it is also seen in `ExternalProject.cmake`.
</blockquote></details>

I would not call that a convention especially.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNNSLoxjIPP2aLW7Tr:bn24lkj) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

You're the expert, but I really don't get it.  How is this not:

```
b = a;
a = c;
fn(c, d);
a = b;
```
</blockquote></details>

`fn` uses a value of a global variable `a` that is different from the value of `a` that is used by the rest of the program.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 8 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/doc/dynamic_analysis_tools.rst, line 39 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFUH0HBywIIOQ5gYu:-KZNP9_wTfVhFSwDwnpw:b-6oofxq) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L39)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

AddressSanitizer (which is why I'm commenting here).

>  If it is anything other than Valgrind, you do not need the MemCheck.

Sounds like the documentation should be corrected here and below, then.

</blockquote></details>

Yes. Thought I did that already. Will fix. There is a bug somewhere in CMake that prevents this from working.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 8 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNm9SHx-EvRPZNcPgw:b-wuf2gp) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
> fn uses a value of a global variable a that is different from the value of a that is used by the rest of the program.

...it does?  I don't see `CMAKE_REQUIRED_LIBRARIES` anywhere else in this file, so I  guess you mean that (some? all?) `check_*_compiler_flag` calls use it.  But I can't find that in the documentation, nor does Googling `CMAKE_REQUIRED_LIBRARIES` provide any useful pointers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 6 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[cmake/tools.cmake, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNCc5aP9l7j3W790IF:-KZNuZ7746Dfr4TXzB7s:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L43)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

I would not call that a convention especially.
</blockquote></details>

Done.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNtWzDqxEuuGqZUql0:b6hofev) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

> fn uses a value of a global variable a that is different from the value of a that is used by the rest of the program.

...it does?  I don't see `CMAKE_REQUIRED_LIBRARIES` anywhere else in this file, so I  guess you mean that (some? all?) `check_*_compiler_flag` calls use it.  But I can't find that in the documentation, nor does Googling `CMAKE_REQUIRED_LIBRARIES` provide any useful pointers.

</blockquote></details>

`check_*_compiler_flag` calls do use it indirectly, yes. To be honest, the best thing to usually do is search the source code of CMake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 6 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZNv45rxAhEGd8PA4pX:b-ut1r42) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

I will look into it.
</blockquote></details>

Was this with Ninja or Make?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 6 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZNwYMdW5k-gWsIWLXL:bfrvbd3) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

`check_*_compiler_flag` calls do use it indirectly, yes. To be honest, the best thing to usually do is search the source code of CMake.
</blockquote></details>

In that case, please have mercy on the next reader and add an explanatory comment about how it's used, and why it needs to be handled this way.

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZNwk-L88cD1Xsw7fPW:b6ors29) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Was this with Ninja or Make?
</blockquote></details>

`ninja`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 6 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZNxDWl4yK1ZOLq9ZBG:b-fqsq31) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

`ninja`
</blockquote></details>

Thought so. Ninja not supporting Fortran always causes issues.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 6 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmake/tools.cmake, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNBD6JhWtivuz1C4vL:-KZO-CuQxfCZR4cQAQBr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L74)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

In that case, please have mercy on the next reader and add an explanatory comment about how it's used, and why it needs to be handled this way.
</blockquote></details>

Done.

---

*[drake/doc/dynamic_analysis_tools.rst, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZNFus7F5wt3kWBsnHx:-KZO-7kPhQ3vylWnOQ-B:b6eogo) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/drake/doc/dynamic_analysis_tools.rst#L65)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Thought so. Ninja not supporting Fortran always causes issues.
</blockquote></details>

Should be fixed, but it is pretty ugly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 6 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmake/tools.cmake, line 280 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-KZND7GH1YEVBZQnvPIa:-KZO-FGgj_KWfYMgdskA:b-ucfkn4) ([raw file](https://github.com/robotlocomotion/drake/blob/97098a76ff4aa6842bbe32bef57fd7ca5a675199/cmake/tools.cmake#L280)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

FYI I've added this to #3873.
</blockquote></details>

Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3920#-:-KZOPE8XZ7KdJ0vmzS4l:bnfp4nl)*
<!-- Sent from Reviewable.io -->
